posts/2023-10-31-23.0.0.11-beta.adoc

```diff
@@ -0,0 +1,330 @@
+
+The method signature of accepting a payload indicates that the acknowledgement strategy for this method is `POST_PROCESSING`. The implementation handles `ack()` and `nack()` calls on the message after the method or chain completes. The upstream data receives the negative acknowledgement with the reason of `IllegalArgumentException`.
+
+=== Emitters
+
+One significant limitation of MicroProfile Reactive Messaging 1.0 was how to integrate imperative code such as RESTful resources and beans using Reactive Messaging which would put and take messages on to a channel according to the `Outgoing` or `Incoming` annotations. In version 3.0, emitters provide a bridge across the two models.
+
+To inject emitters using CDI into a RESTful resources to put messages on to a given channel:
+
+[source,java]
+----
+@Inject
+@Channel(CHANNEL_NAME)
+Emitter<String> emitter;
+
+@POST
+@Path("/payload")
+public CompletionStage<Void> emitPayload(String payload){
+    CompletionStage<Void> cs = emitter.send(payload);
+    return cs;
+}
+
+@POST
+@Path("/message")
+public CompletionStage<Void> emitPayload(String payload){
+    CompletableFuture<Void> ackCf = new CompletableFuture<>();
+    emitter.send(Message.of(payload,
+        () -> {
+            ackCf.complete(null);
+            return CompletableFuture.completedFuture(null);
+        },
+        t -> {
+            ackCf.completeExceptionally(t);
+            return CompletableFuture.completedFuture(null);
+        }));
+    return ackCf;
+}
+----
+
+When defining emitters, you define the type of Object that will be sent as either the payload or the contents of the Message.
+
+If an emitter sends a payload, MicroProfile Reactive Messaging automatically handles the invocation of `ack()` and `nack()` on the message. If, however, the emitter sends a message, the sending code must handle the message being either acked or nacked downstream.
+
+=== Backpressure support
+
+Backpressure support allows for the situation in which messages or payloads are emitted faster than the consumption of the messages. Defining a backpressure strategy allows for an application to define its behaviour in this circumstance. In the following example, the buffer holds up to 300 messages and throws an exception if it is full when a new message is emitted:
+
+[source,java]
+----
+@Inject @Channel("myChannel")
+@OnOverflow(value=OnOverflow.Strategy.BUFFER, bufferSize=300)
+private Emitter<String> emitter;
+
+public void publishMessage() {
+  emitter.send("a");
+  emitter.send("b");
+  emitter.complete();
+}
+----
+
+
+The following backpressure strategies can be defined:
+
+* `BUFFER` - Use a buffer, whose size will be determined by the value of bufferSize if set. Otherwise, the size will be the value of the config property mp.messaging.emitter.default-buffer-size if it exists. Otherwise, 128 will be used. If the buffer is full, an exception will be thrown from the send method. This is the default strategy if no other strategy is defined.
```

Comment:
```suggestion
* `BUFFER` - Use a buffer, with a size determined by the value of bufferSize, if set. Otherwise, the size is the value of the `mp.messaging.emitter.default-buffer-size` MicroProfile Config property,  if it exists. If neither of these values is defined, the default size is 128. If the buffer is full, an exception is thrown from the send method. This is the default strategy if no other strategy is defined.
```